### DISCUSSION TITLE
Does migrateAndValidate migrate always to the latest version?

### DISCUSSION THREAD

**Discussion by MiguelSOliveira** [simolus3/drift#1079] [Q&A]
I'm trying to verify migrations using the test api, like so:

schemaVersion was updated from v8 to v9.

This test fails:

```
test('upgrade from v7 to v8', () async {
    final connection = await verifier.startAt(7);
    final db = MoorDb.connect(connection);
    await db.customStatement('PRAGMA schema_version = 7');
    await verifier.migrateAndValidate(db, 8, validateDropped: true);
  });
```

Output "From" and "to":
"Upgrading v7 to v9"

Because apparently, even though i specify in the `migrateAndValidate` 8 as the target version, it still migrates to v9? Is this the expected behavior?

**Comment by MiguelSOliveira** [comment]
it seems that the problem is here:

```
    await db.executor.ensureOpen(_DelegatingUser(expectedVersion, db));
    final actualSchema = await db.executor.collectSchemaInput();
```

IT uses v9 instead of v8 and I don't understand why.

**Reply by simolus3** [reply]
No, this doesn't look like the expected behavior. `migrateAndValidate` should migrate to the version you're specifying (8 in this case). Is your `onUpgrade` method aware that the newer version might be less than the `schemaVersion` you have in your database? If your migration handler always upgrades to the latest version then that might be what you're seeing?

If you have a small repro that you can share I'd love to take a look at it. Inside this repo we have `extras/migrations_example` that appears to work as intended.

**Reply by MiguelSOliveira** [reply]
Hey Simon, thanks for the answer.

here is my onUpgrade:

```
onUpgrade: (migrator, from, to) async {
          // ignore: avoid_print
          print('Upgrading v$from to v$to');

          if (to < from) {
            // ignore: avoid_print
            print('Trying to downgrade from v$from to v$to');
            throw Exception("Can't downgrade database");
          }

          await customStatement('PRAGMA foreign_keys = OFF');

          await transaction(
            () async {
              for (var i = from; i < to; i++) {
                if (!migrationsMap.containsKey(i)) {
                  // ignore: avoid_print
                  print('No migration available for v$i');
                } else {
                  await migrationsMap[i](migrator, this);
                }
              }
            },
          );

          await customStatement('PRAGMA foreign_keys = ON');
        },
```

When I run the above test it prints "Upgrading v7 to v9" which is incorrect. Can you tell me where the "to" value is specified in the verifier?

**Reply by MiguelSOliveira** [reply]
Also worth noting that on schema_v7 and schema_v8, the new field does not exist (which is why the test is failing, because it expects this new column which was introduced in v9)

**Reply by MiguelSOliveira** [reply]
Which of these is the expectedVersion schema?

```
await db.executor.ensureOpen(_DelegatingUser(expectedVersion, db));
    final actualSchema = await db.executor.collectSchemaInput();

    // Open another connection to instantiate and extract the reference schema.
    final otherConnection = await startAt(expectedVersion);
    await otherConnection.executor.ensureOpen(_DelegatingUser(expectedVersion));
    final referenceSchema = await otherConnection.executor.collectSchemaInput();
    await otherConnection.executor.close();
```

Is it referenceSchema or actualSchema? Both use the expectedVersion?


What version I use in `schemaVersion` is used as the "to" variable...

Ex: when schemaVersion is set arbitrarily to 19 i get:
```
Upgrading v7 to v19
Migrating v7
Migrating v8
No migration available for v9
No migration available for v10
No migration available for v11
No migration available for v12
No migration available for v13
No migration available for v14
No migration available for v15
No migration available for v16
No migration available for v17
No migration available for v18
```

**Reply by MiguelSOliveira** [reply]
Hey @simolus3, I managed to track down the issue.

`await db.customStatement('PRAGMA schema_version = 7');`

I used to need this line, on a previous version because the verifier.startAt wasn't working. Now apparently it isn't needed anymore.. 

Still don't understand why this line would cause everything to break though.

**Reply by davidmartos96** [reply]
I don't thing that schema_version should be used to change the version, as that is used internally by Sqlite.
The times I've needed to force a migration in some of my tests I used the pragma user_version instead.
The schema_version could even cause corruption. Be aware of that.

https://www.sqlite.org/pragma.html#pragma_schema_version

**Reply by kuhnroyal** [reply]
Moor internally uses `user_version` even so the field in the database class is called `schemaVersion`. Don't mess with `schema_version` manually.
